From ab189465509b49f0593d7af4c746a92f324aa2fa Mon Sep 17 00:00:00 2001
From: Damien Thenot <damien.thenot@vates.tech>
Date: Wed, 12 Mar 2025 12:05:12 +0100
Subject: [PATCH] Set RW children of a LVM VDI to coalesce

Children need to be writable so tapdisk can change parent while doing
commit job.

Signed-off-by: Damien Thenot <damien.thenot@vates.tech>
---
 drivers/cleanup.py | 33 +++++++++++++++++++++++++++------
 1 file changed, 27 insertions(+), 6 deletions(-)

diff --git a/drivers/cleanup.py b/drivers/cleanup.py
index 3ca2e0c0..a59b9ff5 100755
--- a/drivers/cleanup.py
+++ b/drivers/cleanup.py
@@ -1408,25 +1408,46 @@ class LVMVDI(VDI):
         if VdiType.isCowImage(self.vdi_type):
             VDI.validate(self, fast)
 
+    def _setChainRw(self) -> List[str]:
+        """
+        Set the readonly LV and children writable.
+        It's needed because the coalesce can be done by tapdisk directly
+        and it will need to write parent information for children.
+        The VDI we want to coalesce into it's parent need to be writable for libqcow coalesce part.
+        Return a list of the LV that were previously readonly to be made RO again after the coalesce.
+        """
+        was_ro = []
+        if self.lvReadonly:
+            self.sr.lvmCache.setReadonly(self.fileName, False)
+            was_ro.append(self.fileName)
+
+        for child in self.children:
+            if child.lvReadonly:
+                self.sr.lvmCache.setReadonly(child.fileName, False)
+                was_ro.append(child.fileName)
+
+        return was_ro
+
+    def _setChainRo(self, was_ro: List[str]) -> None:
+        """Set the list of LV in parameters to readonly"""
+        for lvName in was_ro:
+            self.sr.lvmCache.setReadonly(lvName, True)
+
     @override
     def _doCoalesce(self) -> None:
         """LVMVDI parents must first be activated, inflated, and made writable"""
-        was_RO = False
         try:
             self._activateChain()
             self.sr.lvmCache.setReadonly(self.parent.fileName, False)
             self.parent.validate()
             self.inflateParentForCoalesce()
-            if self.lvReadonly:
-                self.sr.lvmCache.setReadonly(self.fileName, False)
-                was_RO = True
+            was_ro = self._setChainRw()
             VDI._doCoalesce(self)
         finally:
             self.parent._loadInfoSizePhys()
             self.parent.deflate()
             self.sr.lvmCache.setReadonly(self.parent.fileName, True)
-            if was_RO:
-                self.sr.lvmCache.setReadonly(self.fileName, True)
+            self._setChainRo(was_ro)
 
     @override
     def _setParent(self, parent) -> None:
